shell/platform/android/io/flutter/embedding/engine/systemchannels/RestorationChannel.java

```diff
@@ -0,0 +1,144 @@
+   * by the framework until {@code setRestorationData} has been called. It must be set to false if
+   * the engine never calls {@code setRestorationData}. If it has been set to true, but it later
+   * turns out that there is no restoration data, {@code setRestorationData} must be called with
+   * null.
+   */
+  public final boolean waitForRestorationData;
+
+  private MethodChannel channel;
+  private byte[] restorationData;
+  private MethodChannel.Result pendingResult;
+  private boolean engineHasProvidedData = false;
+  private boolean frameworkHasRequestedData = false;
+
+  /** Obtain the most current restoration data that the framework has provided. */
+  public byte[] getRestorationData() {
+    return restorationData;
+  }
+
+  /** Set the restoration data from which the framework will restore its state. */
+  public void setRestorationData(byte[] data) {
+    engineHasProvidedData = true;
+    if (pendingResult != null) {
+      pendingResult.success(data);
+      pendingResult = null;
+      restorationData = data;
+    } else if (frameworkHasRequestedData) {
+      channel.invokeMethod(
+          "push",
+          data,
+          new MethodChannel.Result() {
+            @Override
+            public void success(Object result) {
+              restorationData = data;
+            }
+
+            @Override
+            public void error(String errorCode, String errorMessage, Object errorDetails) {
+              Log.e(
+                  TAG,
+                  "Error "
+                      + errorCode
+                      + " while sending restoration data to framework: "
+                      + errorMessage);
+            }
+
+            @Override
+            public void notImplemented() {
+              // Nothing to do.
+            }
+          });
+    } else {
+      restorationData = data;
+    }
+  }
+
+  /**
+   * Clears the current restoration data.
+   *
+   * <p>This should be called just prior to a hot restart. Otherwise, after the hot restart the
+   * state prior to the hot restart will get restored.
+   */
+  public void clearData() {
+    restorationData = null;
+  }
+
+  private final MethodChannel.MethodCallHandler handler =
+      new MethodChannel.MethodCallHandler() {
+        @Override
+        public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
+          final String method = call.method;
+          final Object args = call.arguments;
+          Log.v(TAG, "Received '" + method + "' message.");
+          switch (method) {
+            case "put":
+              restorationData = (byte[]) args;
+              result.success(null);
+              break;
+            case "get":
+              frameworkHasRequestedData = true;
+              if (engineHasProvidedData || !waitForRestorationData) {
+                result.success(restorationData);
+              } else {
+                pendingResult = result;
```

Comment:
the names are a bit confusing. Can we rename waitForRestorationData to shouldFrameworkWaitForAsyncRestorationData and rename pendingResult to isFrameworkWaitingForRestorationData?